onnxruntime/core/mlas/lib/qgemm_kernel_neon.cpp

```diff
@@ -508,3 +496,477 @@ const MLAS_GEMM_U8X8_DISPATCH MlasGemmU8X8DispatchNeon = {
+            int8x16_t v1 = vreinterpretq_s8_u8(veorq_u8(vld1q_u8(&D[16]), BitFlipVector));
+
+            RowSums0 = vpadalq_s16(RowSums0, vpaddlq_s8(v0));
+            RowSums1 = vpadalq_s16(RowSums1, vpaddlq_s8(v1));
+
+            vst1q_u8(&D[0], vreinterpretq_u8_s8(v0));
+            vst1q_u8(&D[16], vreinterpretq_u8_s8(v1));
+
+            D += 32;
+        }
+
+        RowSums0 = vpaddq_s32(RowSums0, RowSums1);
+        RowSums0 = vpaddq_s32(RowSums0, RowSums0);
+
+        vst1_s32(RowSumBuffer, vget_low_s32(RowSums0));
+        RowSumBuffer += 2;
+
+        A = A + lda * 2;
+    }
+
+    //
+    // Process one row of matrix A.
+    //
+
+    if ((CountM & 1) != 0) {
+
+        const uint8_t* a0 = A;
+
+        size_t k = CountK;
+        int32x4_t RowSums0 = vdupq_n_s32(0);
+
+        while (k >= 16) {
+
+            int8x16_t v0 = vreinterpretq_s8_u8(veorq_u8(vld1q_u8(a0), BitFlipVector));
+            a0 += 16;
+
+            RowSums0 = vpadalq_s16(RowSums0, vpaddlq_s8(v0));
+
+            vst1q_u8(&D[0], vreinterpretq_u8_s8(v0));
+
+            D += 16;
+            k -= 16;
+        }
+
+        if (k > 0) {
+
+            uint8_t* d = D;
+
+            vst1q_u8(&D[0], BitFlipVector);
+
+            while (k > 0) {
+
+                d[0] = *a0++;
+
+                d += 1;
+                k -= 1;
+            }
+
+            int8x16_t v0 = vreinterpretq_s8_u8(veorq_u8(vld1q_u8(&D[0]), BitFlipVector));
+
+            RowSums0 = vpadalq_s16(RowSums0, vpaddlq_s8(v0));
+
+            vst1q_u8(&D[0], vreinterpretq_u8_s8(v0));
+
+            D += 16;
+        }
+
+#if defined(_M_ARM64)
+        // N.B. The workaround of defining a local vaddvq_u32 doesn't work here
+        // as VS2019 added new intrinsics to make the operation work. Also, not
+        // all build environments using VS2019 have the up-to-date arm64_neon.h,
+        // so fallback to pairwise addition.
+        RowSums0 = vpaddq_s32(RowSums0, RowSums0);
+        RowSums0 = vpaddq_s32(RowSums0, RowSums0);
+        vst1q_lane_s32(RowSumBuffer, RowSums0, 0);
+#else
+        *RowSumBuffer = vaddvq_s32(RowSums0);
+#endif
+    }
+}
+
+template<>
+void
+MlasGemmU8X8CopyPackB<MLAS_GEMM_S8S8_KERNEL_NEON>(
+    MLAS_GEMM_S8S8_KERNEL_NEON::PackedBType* D,
+    const uint8_t* B,
+    size_t ldb,
+    size_t CountN,
+    size_t CountK,
+    int32_t* ColumnSumBuffer,
+    bool BIsSigned
+    )
+{
+    MLAS_UNREFERENCED_PARAMETER(BIsSigned);
+
+    while (CountN >= 4) {
+
+        const uint8_t* b = B;
+        size_t k = CountK;
+        int32x4_t ColumnSums0 = vdupq_n_s32(0);
+        int32x4_t ColumnSums1 = vdupq_n_s32(0);
+        int32x4_t ColumnSums2 = vdupq_n_s32(0);
+        int32x4_t ColumnSums3 = vdupq_n_s32(0);
+
+        while (k >= 16) {
+
+            for (size_t nn = 0; nn < 4; nn++) {
+                for (size_t kk = 0; kk < 16; kk++) {
+                    D[nn * 16 + kk] = (b[kk * ldb + nn] );
+                }
+            }
```

Comment:
Is there a vector instruction that speeds it up?

Reply:
future work